airsonic-main/src/main/java/org/airsonic/player/dao/LegacyHsqlDaoHelper.java

```diff
@@ -31,29 +51,226 @@ public void checkpoint() {
+     * @return true if a database migration should be performed
+     */
+    private boolean isHsqldbDatabaseUpgradeNeeded() {
+        // Check the current database version
+        String currentVersion = getHsqldbDatabaseVersion();
+        if (currentVersion == null) {
+            LOG.warn("HSQLDB database version is unknown. Trying to connect anyway, this may upgrade the database.");
+            return false;
+        }
+
+        // Check the database driver version
+        String driverVersion = null;
+        try {
+            Driver driver = DriverManager.getDriver(settingsService.getDefaultJDBCUrl());
+            driverVersion = String.format("%d.%d", driver.getMajorVersion(), driver.getMinorVersion());
+            if (driver.getMajorVersion() != 2) {
+                LOG.warn("HSQLDB database driver version {} is untested. Trying to connect anyway, this may upgrade the database from version {}.", driverVersion, currentVersion);
+                return false;
+            }
+        } catch (SQLException e) {
+            LOG.warn("HSQLDB database driver version cannot be determined. Trying to connect anyway, this may upgrade the database from version {}.", currentVersion, e);
+            return false;
+        }
+
+        // Log what we're about to do and determine if we should perform a controlled update
+        if ("1.8.0".equals(currentVersion) || "1.8.1".equals(currentVersion)) {
+            // If we're on a 1.8.0 or 1.8.1 database and upgrading to 2.x, we're going to handle this manually and check what we're doing.
+            LOG.info("HSQLDB database upgrade needed, from version {} to {}.", currentVersion, driverVersion);
+            return true;
+        } else if (currentVersion.startsWith(driverVersion)) {
+            // If we're already on the same version as the driver, nothing should happen.
+            LOG.info("HSQLDB database upgrade unneeded, already on version {}.", driverVersion);
+            return false;
+        } else if (currentVersion.startsWith("2.")) {
+            // If the database version is 2.x but older than the driver, the upgrade should be relatively painless.
+            LOG.info("HSQLDB database will be silently upgraded from version {} to {}.", currentVersion, driverVersion);
+            return false;
+        } else {
+            // If this happens we're on a completely untested version and we don't know what will happen.
+            LOG.warn("HSQLDB database upgrade from version {} to version {} is untested. Trying to connect anyway, this may upgrade the database.", currentVersion, driverVersion);
+            return false;
+        }
+    }
+
+    /**
+     * Perform a backup of the HSQLDB database, to a timestamped directory.
+     * @return the path to the backup directory
+     */
+    private Path performHsqldbDatabaseBackup() throws IOException {
+
+        String timestamp = new SimpleDateFormat("yyyyMMddHHmmss").format(new Date());
```

Comment:
Maybe use the newer java.time classes here?
```
DateTimeFormatter.ofPattern("xyz").format(Instant.now());
```

Reply:
Thanks for the tip, I'll change that!